### ISSUE TITLE
Lead Conversion Always Defaults Primary Address Type to Home Even if Mapped

### ISSUE THREAD

**Issue by jesslopez-sf** [SalesforceFoundation/NPSP#5538]
If a customer maps a custom Lead field (called Primary Address Type, for example) to the Primary Address Type field on Contact, the Contact's Primary Address Type always defaults to "Home."

This happens because our LD_LeadConvertOverride_CTRL Apex [temporarily disables address management ](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/LD_LeadConvertOverride_CTRL.cls#L413-L417)and then [manually inserts the Address after the fact](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/LD_LeadConvertOverride_CTRL.cls#L474-L483). On Line 481, we call

`
ADDR_Addresses_TDTM.copyAddressStdSObjAddr(con, 'Mailing', addr, null);
`

Since the addrType parameter is null, [we set Address_Type__c to the defaultAddressType value](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/ADDR_Addresses_TDTM.cls#L1133) which is the [default picklist value for the Address_Type field](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/ADDR_Addresses_TDTM.cls#L1306-L1318): (which is "Home").

### Steps to Repeat
1. Create picklist field on Lead object called "Primary Address Type". Add the following values: Home, Work, Other
2. Go to Lead Fields in Object Manager and click the Map Lead Fields button. Map Primary Address Type on Lead to Primary Address Type on Contact
3. Create a Lead with Primary Address Type set to "Work" and a Mailing Address filled in.
4. Click Save and then click the Convert button.
5. The newly created Contact will have a Primary Address Type of "Home"

### Potential Code Fix
Since the Contact is already created with the mapped value in its Primary Address Type field, it should retain it. I mocked up a quick code fix and it worked as I would expect.

[We query for some of the newly created Contact's fields on line 436 of LD_LeadConvertOverride_CTRL.](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/LD_LeadConvertOverride_CTRL.cls#L436)

I added the npe01__Primary_Address_Type__c field to the strSoql variable on line 434 that ultimately builds the fields that we will query for on the newly created Contact.

I then added a String variable to the [Address creation block on lines 476-483](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/LD_LeadConvertOverride_CTRL.cls#L475-L483) and if the Contact’s Primary Address field is not null, use the value from the Contact (which would be the value that was mapped from the Lead). Otherwise, pass the null value. This would then get the default value of the Address Type field, which we do currently.

```
// now create the address object if appropriate
if (CAO_Constants.isHHAccountModel() && !hhAcctAddrDisabled) {
   String primaryAddressType = (con.npe01__Primary_Address_Type__c <> null) ? 
      con.npe01__Primary_Address_Type__c : null;
   Address__c addr = new Address__c();
   addr.Household_Account__c = myAccount.Id;
   addr.Default_Address__c = true;
   addr.Latest_Start_Date__c = system.today();
   addr.Latest_End_Date__c = null;
   ADDR_Addresses_TDTM.copyAddressStdSObjAddr(con, 'Mailing', addr, primaryAddressType);
   UTIL_DMLService.insertRecord(addr);
}
```

I tested the following scenarios and they worked as expected:

- Lead with Primary Address Type set to “Work”; Contact retained Work as Primary Address Type instead of inheriting “Home” from Address Type default.
- Lead with Primary Address Type set to “Home”; Contact retained Home as Primary Address Type.
- Lead with no value in Primary Address Type but a Mailing Address filled in. Contact had Primary Address Type of “Home” because it inherited the default Home value from the Address Type field.

**Comment by jesslopez-sf** [comment]
**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000007hb9sQAA" target="_blank">W-041860</a>

**Comment by jesslopez-sf** [comment]
W-7781161

**Comment by jesslopez-sf** [comment]
This is resolved in NPSP 3.201
